### DISCUSSION TITLE
Synchronising check-in lists

### DISCUSSION THREAD

**Discussion by kiancross** [pretix/pretix#3466] [Ideas] [answered]
# Problem

We have multiple entrances, some of which can only be used after being assigned a hidden add-on product (e.g., accessible entrance).

To achieve this, I think we need a separate check-in list for the special entrances (e.g., configured only to match the 'Accessible' product), with the `Allow checking in add-on tickets by scanning the main ticket` option selected.

However, if this is done, then the same base ticket can be used more than once to gain entrance through the accessible entrance and other entrances, as the check-in lists are not synchronised.

# Proposed solution

My proposal is to introduce 'Check-in list groups', which allow check-in lists to be grouped together. The intersection of a group `A` and `B` will be synchronised. For example, if list `A` has products `x` and `y` and group `B` has products `y` and `z`, then a check-in for product `y` in list `A` will also add a check-in to list `B`.

This will allow for quite complex access control configurations.

There are definitely some edge cases to consider (e.g., what happens if a group is edited to add a new product after check-ins have already started, circular groups, should it be recursive etc.). However, before I get started, I wanted to see what others think. Is there a better solution I haven't thought of? Would you accept a PR like this? Any hints before I get started?

**Comment by raphaelm** [comment, marked as answer]
> However, before I get started, I wanted to see what others think. 

I agree with the problem and we have heard this from customers before.

> Is there a better solution I haven't thought of?

Yes, we do have a different solution in mind. Your solution is powerful, but introduces many edge cases that would be hard to understand and implement correctly. The concept of an "intersection" becomes pretty complex and unintuitive when taking into account the custom check-in logic with arbitrary boolean logic. If anything, I'd personally prefer a "parent-child" relationship between lists instead of arbitrary grouping. I can also imagine funny edge cases in event series.

The solution we plan on implementing, however, is utilizing the already-existing concept of "gates". Devices, i.e. scanners, can belong to gates already. All that is missing really is to use the "gate" as a input value in the boolean logic. Then, you could use the same list on all your entrances with the condition of e.g. ``(gate = main_entrance) || (product = accessible)``

> Would you accept a PR like this?

In general, yes, although probably not with the original "group" idea. However, we can practically only merge changes to check-in that include all necessary changes to (a) pretix itself (b) our Android an Desktop check-in apps (c) our iOS check-in app, which all offer offline evaluation of the logic as well. Therefore, changes to check-in logic are a **lot** of work usually.

We will *very likely* implement the solution I am talking about in the next ~6 months, so if you can wait, waiting will probably also solve your problem ;)

**Reply by kiancross** [reply]
Thanks for the quick reply, @raphaelm.

> The solution we plan on implementing, however, is utilizing the already-existing concept of "gates". Devices, i.e. scanners, can belong to gates already. All that is missing really is to use the "gate" as a input value in the boolean logic. Then, you could use the same list on all your entrances with the condition of e.g. (gate = main_entrance) || (product = accessible)

This makes sense and definitely sounds better than mine; I'm glad I posted here first!

So the following boolean logic could be implemented with this feature?

  - `(gate = priority && (product = queue_jump || product = accessible)) || (gate = main && product = standard)`

I think I probably have some follow-up questions regarding how the `product` selector matches add-on products, but I'll have a play-around first and start a new discussion with these.

> We will very likely implement the solution I am talking about in the next ~6 months, so if you can wait, waiting will probably also solve your problem ;)

I can wait until New Year, so I might need to implement my own temporary solution in the interim! If there is anything I can do to help speed this up (e.g., work on part of it in conjunction with your team), please let me know :)

**Reply by raphaelm** [reply]
> So the following boolean logic could be implemented with this feature?

Yes.

> I think I probably have some follow-up questions regarding how the product selector matches add-on products

For the most part, it does not :)

**Reply by kiancross** [reply]
> For the most part, it does not :)

Understood - I'll put some thoughts in another discussion.

I'm just confirming there is nothing I can do to help with the development of this feature? If there is, please let me know, as I have some free time and would prefer to spend it on something useful rather than a hacky temporary fix!

**Reply by raphaelm** [reply]
> I'm just confirming there is nothing I can do to help with the development of this feature?

If you want to, you can create a PR that introduces the gate as a logic variable. The actual change shouldn't be very involved, but it will require some deep diving into all the code surrounding that logic editor, so it might take a little while to come up with the solution if you haven't seen that part of the code base yet. 

We might not be able to *merge* it before we get around to working on the client part (assuming you don't want to build that, too), but if you only need server-side scanning (i.e. "online mode"), you could just run your system of the PR branch theoretically, which is still better than a hacky temporary fix since it's likely to be merged at *some* point.

That said, us implementing it before new year is very very likely, since one of our biggest customers requires this feature and their next big event is in early January and they'll want to have some time for testing before…

**Reply by kiancross** [reply]
Ok, cool. I've got a couple of other Pretix-related things I'm looking at - I'll take a look at this after if nothing crops up!

**Reply by raphaelm** [reply]
There is now a first implementation in #3521

**Reply by kiancross** [reply]
This is great news! Thanks for letting me know.
